File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/81ddc2ff46c0c32a450ceddca4f0569c.c
#include <stdio.h>
#include <stdbool.h>

typedef long long ll;

ll safe_mod(ll x, ll m) {
    x %= m;
    if (x < 0) x += m;
    return x;
}

struct barrett {
    unsigned int _m;
    unsigned long long im;
};

barrett create_barrett(unsigned int m) {
    barrett b;
    b._m = m;
    b.im = (unsigned long long)(-1) / m + 1;
    return b;
}

unsigned int umod(barrett b) {
    return b._m;
}

unsigned int mul(barrett b, unsigned int a, unsigned int b) {
    unsigned long long z = a;
    z *= b;
    unsigned long long x = (unsigned long long)(((unsigned __int128)(z)*b.im) >> 64);
    unsigned int v = (unsigned int)(z - x * b._m);
    if (b._m <= v) v += b._m;
    return v;
}

ll id4(ll x, ll n, int m) {
    if (m == 1) return 0;
    unsigned int _m = (unsigned int)(m);
    unsigned long long r = 1;
    unsigned long long y = safe_mod(x, m);
    while (n) {
        if (n & 1) r = (r * y) % _m;
        y = (y * y) % _m;
        n >>= 1;
    }
    return r;
}

bool id1(int n) {
    if (n <= 1) return false;
    if (n == 2 || n == 7 || n == 61) return true;
    if (n % 2 == 0) return false;
    ll d = n - 1;
    while (d % 2 == 0) d /= 2;
    ll bases[3] = {2, 7, 61};
    for (int i = 0; i < 3; i++) {
        ll a = bases[i];
        ll t = d;
        ll y = id4(a, t, n);
        while (t != n - 1 && y != 1 && y != n - 1) {
            y = y * y % n;
            t <<= 1;
        }
        if (y != n - 1 && t % 2 == 0) {
            return false;
        }
    }
    return true;
}

typedef struct {
    ll first;
    ll second;
} pair;

pair id2(ll a, ll b) {
    a = safe_mod(a, b);
    if (a == 0) {
        pair p;
        p.first = b;
        p.second = 0;
        return p;
    }
    ll s = b, t = a;
    ll m0 = 0, m1 = 1;
    while (t) {
        ll u = s / t;
        s -= t * u;
        m0 -= m1 * u;
        ll tmp = s;
        s = t;
        t = tmp;
        tmp = m0;
        m0 = m1;
        m1 = tmp;
    }
    if (m0 < 0) m0 += b / s;
    pair p;
    p.first = s;
    p.second = m0;
    return p;
}

int id3(int m) {
    if (m == 2) return 1;
    if (m == 167772161) return 3;
    if (m == 469762049) return 3;
    if (m == 754974721) return 11;
    if (m == 998244353) return 3;
    int divs[20] = {0};
    divs[0] = 2;
    int cnt = 1;
    int x = (m - 1) / 2;
    while (x % 2 == 0) x /= 2;
    for (int i = 3; (long long)(i)*i <= x; i += 2) {
        if (x % i == 0) {
            divs[cnt++] = i;
            while (x % i == 0) {
                x /= i;
            }
        }
    }
    if (x > 1) {
        divs[cnt++] = x;
    }
    for (int g = 2;; g++) {
        bool ok = true;
        for (int i = 0; i < cnt; i++) {
            if (id4(g, (m - 1) / divs[i], m) == 1) {
                ok = false;
                break;
            }
        }
        if (ok) return g;
    }
}

typedef struct {
    int m;
} id0;

id0 create_id0(int m) {
    id0 id;
    id.m = m;
    return id;
}

int mod(id0 id) {
    return id.m;
}

ll raw(id0 id, int v) {
    return (ll)v;
}

id0 create_id0_default() {
    return create_id0(0);
}

id0 create_id0_from_bool(bool v) {
    return create_id0((int)v);
}

unsigned int val(id0 id, ll x) {
    return (unsigned int)x;
}

id0 operator_plus_plus(id0 id) {
    return id;
}

id0 operator_minus_minus(id0 id) {
    return id;
}

id0 operator_plus_plus_post(id0 id) {
    return id;
}

id0 operator_minus_minus_post(id0 id) {
    return id;
}

id0 operator_plus_equal(id0 id, id0 rhs) {
    id.m += rhs.m;
    if (id.m >= umod(id)) id.m -= umod(id);
    return id;
}

id0 operator_minus_equal(id0 id, id0 rhs) {
    id.m -= rhs.m;
    if (id.m >= umod(id)) id.m += umod(id);
    return id;
}

id0 operator_star_equal(id0 id, id0 rhs) {
    unsigned long long z = id.m;
    z *= rhs.m;
    id.m = (unsigned int)(z % umod(id));
    return id;
}

id0 operator_slash_equal(id0 id, id0 rhs) {
    return id;
}

id0 operator_plus(id0 lhs, id0 rhs) {
    return operator_plus_equal(lhs, rhs);
}

id0 operator_minus(id0 lhs, id0 rhs) {
    return operator_minus_equal(lhs, rhs);
}

id0 operator_star(id0 lhs, id0 rhs) {
    return operator_star_equal(lhs, rhs);
}

id0 operator_slash(id0 lhs, id0 rhs) {
    return operator_slash_equal(lhs, rhs);
}

bool operator_equal_equal(id0 lhs, id0 rhs) {
    return lhs.m == rhs.m;
}

bool operator_exclaim_equal(id0 lhs, id0 rhs) {
    return lhs.m != rhs.m;
}

typedef struct {
    int m;
} id5;

id5 create_id5(int m) {
    id5 id;
    id.m = m;
    return id;
}

int mod(id5 id) {
    return id.m;
}

void set_mod(id5 id, int m) {
    id.m = m;
}

ll raw(id5 id, int v) {
    return (ll)v;
}

id5 create_id5_default() {
    return create_id5(0);
}

id5 create_id5_from_bool(bool v) {
    return create_id5((int)v);
}

unsigned int val(id5 id, ll x) {
    return (unsigned int)x;
}

id5 operator_plus_plus(id5 id) {
    return id;
}

id5 operator_minus_minus(id5 id) {
    return id;
}

id5 operator_plus_plus_post(id5 id) {
    return id;
}

id5 operator_minus_minus_post(id5 id) {
    return id;
}

id5 operator_plus_equal(id5 id, id5 rhs) {
    id.m += rhs.m;
    if (id.m >= umod(id)) id.m -= umod(id);
    return id;
}

id5 operator_minus_equal(id5 id, id5 rhs) {
    id.m += mod(id) - rhs.m;
    if (id.m >= umod(id)) id.m -= umod(id);
    return id;
}

id5 operator_star_equal(id5 id, id5 rhs) {
    id.m = mul(id, id.m, rhs.m);
    return id;
}

id5 operator_slash_equal(id5 id, id5 rhs) {
    return id;
}

id5 operator_plus(id5 lhs, id5 rhs) {
    return operator_plus_equal(lhs, rhs);
}

id5 operator_minus(id5 lhs, id5 rhs) {
    return operator_minus_equal(lhs, rhs);
}

id5 operator_star(id5 lhs, id5 rhs) {
    return operator_star_equal(lhs, rhs);
}

id5 operator_slash(id5 lhs, id5 rhs) {
    return operator_slash_equal(lhs, rhs);
}

bool operator_equal_equal(id5 lhs, id5 rhs) {
    return lhs.m == rhs.m;
}

bool operator_exclaim_equal(id5 lhs, id5 rhs) {
    return lhs.m != rhs.m;
}

typedef id0 mint;

mint pow(mint x, ll n) {
    mint r = create_id0(1);
    while (n) {
        if (n & 1) r = r * x;
        x = x * x;
        n >>= 1;
    }
    return r;
}

mint inv(mint x) {
    return pow(x, mod(x) - 2);
}

void solve() {
    int n, m;
    scanf("%d%d", &n, &m);

    mint::set_mod(m);

    mint dp[n + 1];
    mint p[n + 1];

    dp[n] = create_id0(1);
    p[n] = create_id0(1);
    for (int i = n - 1; i >= 1; i--) {
        dp[i] += p[i + 1];

        for (int z = 2; 1U * z * i <= n; z++) {
            int r = i * z + z;
            int l = 1U * z * i;
            dp[i] += p[l];
            if (r <= n) dp[i] -= p[r];
        }

        p[i] = p[i + 1] + dp[i];
    }

    printf("%u\n", val(dp[1]));
}

int main() {
    int _t = 1;

    while (_t--) solve();
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/81ddc2ff46c0c32a450ceddca4f0569c.py
class Barrett:
    def __init__(self, m):
        self._m = m
        self.im = (-1 // m) + 1

    def umod(self):
        return self._m

    def mul(self, a, b):
        z = a * b
        x = (z * self.im) >> 64
        v = z - x * self._m
        if self._m <= v:
            v += self._m
        return v


def safe_mod(x, m):
    x %= m
    if x < 0:
        x += m
    return x


def id4(x, n, m):
    if m == 1:
        return 0
    _m = m
    r = 1
    y = safe_mod(x, m)
    while n:
        if n & 1:
            r = (r * y) % _m
        y = (y * y) % _m
        n >>= 1
    return r


def id1(n):
    if n <= 1:
        return False
    if n in (2, 7, 61):
        return True
    if n % 2 == 0:
        return False
    d = n - 1
    while d % 2 == 0:
        d //= 2
    bases = [2, 7, 61]
    for a in bases:
        t = d
        y = id4(a, t, n)
        while t != n - 1 and y != 1 and y != n - 1:
            y = (y * y) % n
            t <<= 1
        if y != n - 1 and t % 2 == 0:
            return False
    return True


def id2(a, b):
    a = safe_mod(a, b)
    if a == 0:
        return (b, 0)

    s, t = b, a
    m0, m1 = 0, 1

    while t:
        u = s // t
        s -= t * u
        m0 -= m1 * u

        s, t = t, s
        m0, m1 = m1, m0

    if m0 < 0:
        m0 += b // s
    return (s, m0)


def id3(m):
    if m == 2:
        return 1
    if m in (167772161, 469762049, 998244353):
        return 3
    if m == 754974721:
        return 11
    divs = [2]
    cnt = 1
    x = (m - 1) // 2
    while x % 2 == 0:
        x //= 2
    for i in range(3, int(x**0.5) + 1, 2):
        if x % i == 0:
            divs.append(i)
            while x % i == 0:
                x //= i
            cnt += 1
    if x > 1:
        divs.append(x)
        cnt += 1
    for g in range(2, 100):
        ok = True
        for i in range(cnt):
            if id4(g, (m - 1) // divs[i], m) == 1:
                ok = False
                break
        if ok:
            return g
    return -1


class ModInt:
    def __init__(self, v=0, m=998244353):
        self._v = v % m
        self._m = m

    def val(self):
        return self._v

    def __add__(self, other):
        return ModInt((self._v + other._v) % self._m)

    def __sub__(self, other):
        return ModInt((self._v - other._v) % self._m)

    def __mul__(self, other):
        return ModInt((self._v * other._v) % self._m)

    def __truediv__(self, other):
        return self * other.inv()

    def inv(self):
        eg = id2(self._v, self._m)
        assert eg[0] == 1
        return ModInt(eg[1], self._m)

    def pow(self, n):
        assert n >= 0
        x = self
        r = ModInt(1, self._m)
        while n:
            if n & 1:
                r *= x
            x *= x
            n >>= 1
        return r


def solve():
    n, m = map(int, input().split())
    ModInt._m = m

    dp = [ModInt() for _ in range(n + 1)]
    p = [ModInt() for _ in range(n + 1)]

    dp[n] = ModInt(1)
    p[n] = ModInt(1)
    for i in range(n - 1, 0, -1):
        dp[i] += p[i + 1]

        z = 2
        while z * i <= n:
            r = i * z + z
            l = z * i
            dp[i] += p[l]
            if r <= n:
                dp[i] -= p[r]
            z += 1

        p[i] = p[i + 1] + dp[i]

    print(dp[1].val())


if __name__ == "__main__":
    import sys
    input = sys.stdin.read
    solve()
